### ISSUE TITLE
Flaky ConfigurationErrorsException

### ISSUE THREAD

**Issue by ishatalkin** [aspnet/MicrosoftConfigurationBuilders#121]
Sometimes whern I run my application I get ConfigurationErrorsException:
The configBuilder 'envParams' failed while processing raw xml.: Error in Configuration Builder 'envParams'::GetValue(***). 

However the value *** is presented in json-file, and if I rerun my app, there will be no error.

Any ideas what is the reason of the error?

I use Microsoft.Configuration.ConfigurationBuilders.Json, version 2.0.0.

Му app.config is as follows:
```xml
  <configSections>
     <section name="configBuilders" type="System.Configuration.ConfigurationBuildersSection, System.Configuration, Version=4.0.0.0, Culture=neutral, publicKeyToken=b03f5f7f11d50a3a" requirePermission="false" />
  </configSections>
  <configBuilders >
    <builders>
      <add name="envParams" mode="Expand" jsonFile="~\bin\App_Data\envParams.Dev.json" optional="false" 
           stripPrefix="true" type="Microsoft.Configuration.ConfigurationBuilders.SimpleJsonConfigBuilder, Microsoft.Configuration.ConfigurationBuilders.Json, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35" />
    </builders>
  </configBuilders>
```

Stack trace:

```
[The configBuilder 'envParams' failed while processing raw xml.: Error in Configuration Builder 'envParams'::GetValue(***) 
  ----> System.Configuration.ConfigurationErrorsException : The configBuilder 'envParams' failed while processing raw xml.: Error in Configuration Builder 'envParams'::GetValue(***) 
  ----> System.Exception : Error in Configuration Builder 'envParams'::GetValue(***)
  ----> System.Collections.Generic.KeyNotFoundException
--ConfigurationErrorsException
   в System.Configuration.BaseConfigurationRecord.EvaluateOne(String[] keys, SectionInput input, Boolean isTrusted, FactoryRecord factoryRecord, SectionRecord sectionRecord, Object parentResult)
   в System.Configuration.BaseConfigurationRecord.Evaluate(FactoryRecord factoryRecord, SectionRecord sectionRecord, Object parentResult, Boolean getLkg, Boolean getRuntimeObject, Object& result, Object& resultRuntimeObject)
   в System.Configuration.BaseConfigurationRecord.GetSectionRecursive(String configKey, Boolean getLkg, Boolean checkPermission, Boolean getRuntimeObject, Boolean requestIsHere, Object& result, Object& resultRuntimeObject)
   в System.Configuration.BaseConfigurationRecord.GetSectionRecursive(String configKey, Boolean getLkg, Boolean checkPermission, Boolean getRuntimeObject, Boolean requestIsHere, Object& result, Object& resultRuntimeObject)
   в System.Configuration.BaseConfigurationRecord.GetSectionRecursive(String configKey, Boolean getLkg, Boolean checkPermission, Boolean getRuntimeObject, Boolean requestIsHere, Object& result, Object& resultRuntimeObject)
   в System.Configuration.BaseConfigurationRecord.GetSection(String configKey)
   в System.Configuration.ConfigurationManager.GetSection(String sectionName)
--Exception
   в Microsoft.Configuration.ConfigurationBuilders.KeyValueConfigBuilder.GetValueInternal(String key)
   в Microsoft.Configuration.ConfigurationBuilders.KeyValueConfigBuilder.<ExpandTokens>b__47_0(Match m)
   в System.Text.RegularExpressions.RegexReplacement.Replace(MatchEvaluator evaluator, Regex regex, String input, Int32 count, Int32 startat)
   в System.Text.RegularExpressions.Regex.Replace(String input, MatchEvaluator evaluator)
   в System.Text.RegularExpressions.Regex.Replace(String input, String pattern, MatchEvaluator evaluator)
   в Microsoft.Configuration.ConfigurationBuilders.KeyValueConfigBuilder.ExpandTokens(XmlNode rawXml)
   в System.Configuration.BaseConfigurationRecord.ProcessRawXml(ConfigXmlReader reader, ConfigurationBuilder configBuilder)
--KeyNotFoundException
   в System.ThrowHelper.ThrowKeyNotFoundException()
   в System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   в Microsoft.Configuration.ConfigurationBuilders.SimpleJsonConfigBuilder.GetCurrentDictionary()
   в Microsoft.Configuration.ConfigurationBuilders.SimpleJsonConfigBuilder.GetValue(String key)
   в Microsoft.Configuration.ConfigurationBuilders.KeyValueConfigBuilder.GetValueInternal(String key)
```

**Comment by ishatalkin** [comment]
Any suggestions?

**Comment by ishatalkin** [comment]
Ok. I found the reason of this error.

The reason is that sometimes in multithreading environment JSON builder is used not fully initialized. Assume the folowing scenario.

1) Thread 1 begins using of SimpleJsonConfigBuilder. Method KeyValueConfigBuilder.EnsureInitialized is invoked. It checks _lazyInitialized field. The object is not initialized yet. Ok, SimpleJsonConfigBuilder.LazyInitialize is invoked, which invokes LazyInitialize of base class.
2) LazyInitialize of base class does its work and sets _lazyInitialized to true. Then it gives control back to SimpleJsonConfigBuilder.LazyInitialize.
3) Assume at this moment starts Thread 2, which uses the same json builder. Method KeyValueConfigBuilder.EnsureInitialized is invoked. It checks _lazyInitialized field. The computer thinks object is initialized. Ok, we can use it.
4) Thread 2 tries to do some stuff with object, but ooops! It's not fully initialized, as Thread 1 hasn't initialized the object in SimpleJsonConfigBuilder.LazyInitialize yet.

**Comment by ishatalkin** [comment]
Opened pull request #134

**Comment by StephenMolloy** [comment]
Fixed by #134

**Comment by StephenMolloy** [comment]
Fixed by #134 and #164
